examples/step-17/step-17.cc

```diff
@@ -362,51 +365,50 @@ namespace Step17
-    // triangulation and DoFHandler objects, we will simply store
-    // <i>all</i> constraints on each process; again, this will not
-    // scale, but we show in step-40 how one can work around this by
-    // only storing on each MPI process the constraints for degrees of
-    // freedom that actually matter on this particular process.
-    //
-    // Since PETSc handles the
-    // sparsity pattern internally to the matrix, there is no need to set up
-    // an independent sparsity pattern here, and to condense it for
-    // constraints, as we have done in all other example programs.
+    // We need to initialize the objects denoting hanging node constraints for
+    // the present grid. As with the triangulation and DoFHandler objects, we
+    // will simply store <i>all</i> constraints on each process; again, this
+    // will not scale, but we show in step-40 how one can work around this by
+    // only storing on each MPI process the constraints for degrees of freedom
+    // that actually matter on this particular process.
     hanging_node_constraints.clear ();
     DoFTools::make_hanging_node_constraints (dof_handler,
                                              hanging_node_constraints);
     hanging_node_constraints.close ();
+
+    // Now we create the sparsity pattern for the system matrix. Note that we
+    // again compute and store all constraints and not only the ones relevant
+    // to this process (see step-18 or step-40 for a more efficient way to
+    // handle this).
+    DynamicSparsityPattern dsp(dof_handler.n_dofs(), dof_handler.n_dofs());
+    DoFTools::make_sparsity_pattern (dof_handler, dsp,
+                                     hanging_node_constraints,
+                                     false);
+
+    // Now we determine the set of locally owned DoFs and use that to
+    // initialize parallel vectors and matrix. Since the matrix and vectors
+    // need to work in parallel, we have to pass them an MPI communication
+    // object, as well as information about the partitioning contained in the
+    // IndexSet @p locally_owned_dofs.  The IndexSet contains information
+    // about the global size (the <i>total</i> number of degrees of freedom)
+    // and also what subset of rows is to be stored locally.  Note that the
+    // system matrix needs that partitioning information for the rows and
+    // columns. For square matrices, as it is the case here, the columns
+    // should be partitioned in the same way as the rows, but in the case of
+    // rectangular matrices one has to partition the columns in the same way
+    // as vectors are partitioned with which the matrix is multiplied, while
+    // rows have to partitioned in the same way as destination vectors of
+    // matrix-vector multiplications:
+    std::vector<IndexSet> locally_owned_dofs_per_proc = DoFTools::locally_owned_dofs_per_subdomain(dof_handler);
+    IndexSet locally_owned_dofs = locally_owned_dofs_per_proc[this_mpi_process];
```

Comment:
In fact, is there no way to query the number of locally owned dofs without having to count dofs owned by other processors? It seems wasteful to compute the whole vector, and then only use one element.